### PR TITLE
Add typesafe auditlog config

### PR DESCRIPTION
**Pull request by anuraaga**
I am looking at removing `types.Config` because it is a very implicit type and has awkwardness in using it, with allocations related to defaultValue and `interface{}`. We can use concrete types instead for all our uses as far as I can tell.

This starts by creating an auditlog config type, which needs to be public as its part of the auditlog plugin mechanism. Then it can be removed by adding an internal type for containing parser state/options which is its only other usage right now.

### CODE REVIEW

**Comment by anuraaga** [comment]
@jcchavezs We have the ability to register formatters or writers so it's some sort of pluggability, which is what I mean. Anyways, need to get this in to continue to make some more breaking changes

**Comment by jcchavezs** [comment]
Go for it LGTM

**Comment by anuraaga** [comment]
@jcchavezs Can't merge without approval for this repo :P

**Review by jcchavezs — APPROVED** [review]

**Pull request merged**
